jax/interpreters/pxla.py

```diff
@@ -721,38 +734,42 @@ def dynamic_fun(dummy, *args):
                  for d in xb.local_devices(host_id)]
     else:
       devices = xb.get_backend(backend).get_default_device_assignment(
-          num_global_replicas)
+          num_global_replicas, num_partitions)
   else:
-    if num_local_replicas != len(local_devices):
+    if num_local_shards != len(local_devices):
       local_devices_str = ", ".join(map(str, local_devices))
       raise ValueError(
           "Leading axis size of input to pmapped function must equal the "
           "number of local devices passed to pmap. Got axis_size=%d, "
           "num_local_devices=%d.\n(Local devices passed to pmap: %s)"
           % (axis_size, len(local_devices), local_devices_str))
-    if num_global_replicas != len(devices):
-      raise ValueError("compiling computation that requires %s replicas, "
-                       "but %s devices were specified"
-                       % (num_global_replicas, len(devices)))
+    if num_global_shards != len(devices):
+      raise ValueError("compiling computation that creates %s shards, "
+                       "but %s devices were specified" %
+                       (num_global_shards, len(devices)))
 
-  device_assignment = tuple(d.id for d in devices)
+  device_assignment = tree_map(lambda d: d.id, devices)
```

Comment:
Isn't `devices` a list or a tuple? In that case I think that `tree_map` is overkill, and `tuple([d.id for d in devices])` may be enough.

Reply:
I think it's now a list of lists?

Reply:
It may be either 1D or 2D, depending on if we got a 1D `devices` list/tuple or a 2D assignment from `get_default_device_assignment`. I added a comment.